lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -363,211 +642,256 @@ runSelectionStep lens s
+                -- construct the change.
+                unsafeNaturalToCoin $ distance
+                    (coinToNatural excessCoin)
+                    (coinToNatural requiredCost + totalMinCoinValue)
+            }
+      where
+        totalMinCoinValue =
+            F.sum $ (coinToNatural . minCoinValueFor) <$> change
+
+    outputTokens :: NonEmpty TokenMap
+    outputTokens = view #tokens <$> outputBundles
+
+    outputCoins :: NonEmpty Coin
+    outputCoins = view #coin <$> outputBundles
+
     totalInputValue :: TokenBundle
-    totalInputValue = F.fold inputBundles
+    totalInputValue = TokenBundle.add
+        (F.fold inputBundles)
+        (maybe TokenBundle.empty TokenBundle.fromCoin mExtraCoinSource)
 
     totalOutputValue :: TokenBundle
     totalOutputValue = F.fold outputBundles
 
-    totalOutputCoinValue :: Coin
-    totalOutputCoinValue = TokenBundle.getCoin totalOutputValue
-
-    excess :: TokenBundle
-    excess =
-        -- The following subtraction is safe, as we have already checked that
-        -- the total input value is greater than the total output value:
-        totalInputValue `TokenBundle.unsafeSubtract` totalOutputValue
-
-    change :: NonEmpty TokenBundle
-    change
-        = TokenBundle.fromCoin <$> changeForCoin
-        & NE.zipWith (<>) (TokenBundle.fromTokenMap <$> changeForPaymentAssets)
-        -- Here we sort in ascending order of coin value so that surplus assets
-        -- and coin values are in increasing order in the resulting bundles:
-        & NE.sortWith TokenBundle.getCoin
-        & NE.zipWith (<>) (TokenBundle.fromTokenMap <$> changeForSurplusAssets)
-
-    -- Change for the excess coin quantity.
-    changeForCoin :: NonEmpty Coin
-    changeForCoin = makeChangeForCoin
-        (TokenBundle.getCoin excess)
-        (TokenBundle.getCoin <$> outputBundles)
-
-    -- Change for excess quantities of assets included in outputs.
-    changeForPaymentAssets :: NonEmpty TokenMap
-    changeForPaymentAssets = makeChangeForPaymentAssets
-        (TokenMap.filter (`Set.member` paymentAssetIds) (view #tokens excess))
-        (view #tokens <$> outputBundles)
-
-    -- Change for excess quantities of assets NOT included in outputs.
-    changeForSurplusAssets :: NonEmpty TokenMap
-    changeForSurplusAssets =
-        makeChangeForSurplusAssets surplusAssetsGrouped outputBundles
-      where
-        surplusAssetsGrouped :: Map AssetId (NonEmpty TokenQuantity)
-        surplusAssetsGrouped = groupByKey surplusAssets
-
-        surplusAssets :: [(AssetId, TokenQuantity)]
-        surplusAssets = filter
-            ((`Set.notMember` paymentAssetIds) . fst)
-            (TokenMap.toFlatList . view #tokens =<< NE.toList inputBundles)
-
     -- Identifiers of assets included in outputs.
-    paymentAssetIds :: Set AssetId
-    paymentAssetIds = TokenBundle.getAssets totalOutputValue
+    knownAssetIds :: Set AssetId
+    knownAssetIds = TokenBundle.getAssets totalOutputValue
```

Comment:
I must admit I'm a bit curious as to the reason for the change in terminology here!

I think the term `paymentAssetIds` is actually very clear, as these assets correspond to assets found in the payments/outputs.

But it's not clear to me in what sense a token can be thought of as "unknown". I think this term is potentially a bit confusing.

Reply:
I found `payment` equally confusing though, for the payment is actually the whole transaction. Perhaps, one better way to describe those are:

- userSpecifiedAssets
- changeAssets

What do you think?

Reply:
I think it's important that the two terms are obvious antonyms of each other.

So this would definitely work:

- `userSpecifiedAssets`
- `nonUserSpecifiedAssets`

But perhaps this is even shorter and clearer:

- `outputAssets`
- `nonOutputAssets`

The last option makes it very obvious that the `outputAssets` were those assets actually found in the outputs. (And it would make it consistent with the comments, as those refer to "assets found in the outputs" and "assets not found in the original outputs")

Reply:
I don't like `outputAssets` for the same reasons I don't like `paymentAssets`, change outputs are outputs too :sweat_smile: ... So I think this settles it on `userSpecifiedAssets` / `nonUserSpecifiedAssets` 

I'll make the change :+1:

Reply:
> So I think this settles it on userSpecifiedAssets / nonUserSpecifiedAssets

I agree that the combination of "user-specified" and "non-user-specified" definitely works here. I think it will also work well in any public facing documentation we write about the algorithms.

I also think it is much clearer than "user-defined" (as this could be interpreted as meaning that the user had defined their own assets, which is somewhat ambiguous).

I added some further clarifications to the documentation and comments in 5dfc0d17454059271a7e9ff8bd37b8acee7a3256.